onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc

```diff
@@ -409,27 +432,60 @@ class QLinearConvOpTester {
           float weight_scale = W_.scale_[(W_.scale_.size() == 1) ? 0 : channel_index];
           float requantize_scale = (X_.scale_[0] * weight_scale) / output_scale_;
 
-          for (int64_t oh = 0; oh < output_h; oh++) {
-            for (int64_t ow = 0; ow < output_w; ow++) {
+          if (kernel_rank == 2) {
+            for (int64_t oh = 0; oh < output_h; oh++) {
+              for (int64_t ow = 0; ow < output_w; ow++) {
+                int32_t sum = bias;
+                const T1* input_image = Xdata;
+                const T2* weight_data = weight_row;
+                for (int64_t ic = 0; ic < group_input_channels; ic++) {
+                  for (int64_t kh = 0; kh < kernel_h; kh++) {
+                    int64_t ih = kh * dilation_h + oh * stride_h - pad_t;
+                    for (int64_t kw = 0; kw < kernel_w; kw++) {
+                      int64_t iw = kw * dilation_w + ow * stride_w - pad_l;
+                      int32_t w_value = static_cast<int32_t>(*weight_data++);
+                      if (static_cast<uint64_t>(ih) < static_cast<uint64_t>(input_h) &&
+                          static_cast<uint64_t>(iw) < static_cast<uint64_t>(input_w)) {
+                        int32_t x_value = static_cast<int32_t>(input_image[ih * input_w + iw]) - X_zero_point;
+                        sum += x_value * w_value;
+                      }
+                    }
+                  }
+                  input_image += input_image_size;
+                }
+                *Ydata++ = RequantizeOutput<T1>(sum, requantize_scale, requantize_values);
+              }
+            }
+          } else {
+            std::vector<int64_t> d_output(kernel_rank, 0);
+            std::vector<int64_t> d_kernel(kernel_rank, 0);
+            for (bool has_next_out = true; has_next_out;) {
               int32_t sum = bias;
               const T1* input_image = Xdata;
               const T2* weight_data = weight_row;
               for (int64_t ic = 0; ic < group_input_channels; ic++) {
-                for (int64_t kh = 0; kh < kernel_h; kh++) {
-                  int64_t ih = kh * dilation_h + oh * stride_h - pad_t;
-                  for (int64_t kw = 0; kw < kernel_w; kw++) {
-                    int64_t iw = kw * dilation_w + ow * stride_w - pad_l;
-                    int32_t w_value = static_cast<int32_t>(*weight_data++);
-                    if (static_cast<uint64_t>(ih) < static_cast<uint64_t>(input_h) &&
-                        static_cast<uint64_t>(iw) < static_cast<uint64_t>(input_w)) {
-                      int32_t x_value = static_cast<int32_t>(input_image[ih * input_w + iw]) - X_zero_point;
-                      sum += x_value * w_value;
-                    }
+                for (bool has_next_kernel = true; has_next_kernel;) {
```

Comment:
should this just be a do {} while (NextPosition...) loop and get rid of has_next_kernel?

Reply:
Yeah!